File: src/main/java/org/example/bamsemats/dotatranslator/GoogleOAuthService.java
package org.example.bamsemats.dotatranslator;

import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.client.util.store.FileDataStoreFactory;
import com.google.auth.oauth2.AccessToken;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.security.GeneralSecurityException;
import java.util.Collections;
import java.util.List;

public class GoogleOAuthService {

    private static final String APPLICATION_NAME = "Dota2ChatTranslator";
    private static final JsonFactory JSON_FACTORY = GsonFactory.getDefaultInstance();
    private static final List<String> SCOPES = Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"); // Broad scope for Vision and Translate
    private static final String TOKENS_DIRECTORY_PATH = "tokens"; // Directory to store user's refresh token
    private static final String CLIENT_SECRET_FILE = "/client_secret.json"; // Path to client_secret.json in resources

    private HttpTransport HTTP_TRANSPORT;
    private FileDataStoreFactory DATA_STORE_FACTORY;
    private GoogleClientSecrets clientSecrets; // Store client secrets
    private String clientId;
    private String clientSecret;


    public GoogleOAuthService() throws GeneralSecurityException, IOException {
        HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
        DATA_STORE_FACTORY = new FileDataStoreFactory(new File(TOKENS_DIRECTORY_PATH));

        // Load client secrets in constructor
        InputStream in = GoogleOAuthService.class.getResourceAsStream(CLIENT_SECRET_FILE);
        if (in == null) {
            throw new IOException("Client secret file " + CLIENT_SECRET_FILE + " not found in resources.");
        }
        clientSecrets = GoogleClientSecrets.load(JSON_FACTORY, new InputStreamReader(in));
        
        // Store client ID and secret
        if (clientSecrets.getDetails() != null) {
            this.clientId = clientSecrets.getDetails().getClientId();
            this.clientSecret = clientSecrets.getDetails().getClientSecret();
        } else {
            throw new IOException("Client secret file " + CLIENT_SECRET_FILE + " is missing client ID or client secret details.");
        }
    }

    /**
     * Authorizes the user and returns a Credential object.
     * If a refresh token is available, it will be used to refresh the access token.
     * Otherwise, it will initiate a new authorization flow, opening a browser window.
     * @return Credential object containing access token, refresh token, etc.
     * @throws IOException
     * @throws GeneralSecurityException
     */
    public Credential authorize() throws IOException, GeneralSecurityException {
        // Build flow and trigger user authorization request.
        GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(
                HTTP_TRANSPORT, JSON_FACTORY, clientSecrets, SCOPES) // Use stored clientSecrets
                .setDataStoreFactory(DATA_STORE_FACTORY)
                .setAccessType("offline") // Required to get a refresh token
                .build();

        // Authorize the user. This will open a browser window if needed.
        // It will also handle token storage and refreshing.
        Credential credential = new AuthorizationCodeInstalledApp(flow, new LocalServerReceiver()).authorize("user");
        
        System.out.println("Credentials saved to " + TOKENS_DIRECTORY_PATH);
        return credential;
    }

    /**
     * Retrieves the last known authorized credential without re-authorizing the user.
     * Useful for checking if a user is already authenticated.
     * @return Credential object if found, null otherwise.
     * @throws IOException
     * @throws GeneralSecurityException
     */
    public Credential getStoredCredential() throws IOException, GeneralSecurityException {
        GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(
                HTTP_TRANSPORT, JSON_FACTORY, clientSecrets, SCOPES) // Use stored clientSecrets
                .setDataStoreFactory(DATA_STORE_FACTORY)
                .setAccessType("offline")
                .build();
        
        return flow.loadCredential("user");
    }

    public String getClientId() {
        return clientId;
    }

    public String getClientSecret() {
        return clientSecret;
    }
}

File: google_oauth_service.py
import os
import json
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials

# Scopes required for Google Cloud Vision and Translation APIs
SCOPES = ['https://www.googleapis.com/auth/cloud-platform']

# Path to the client secrets file
CLIENT_SECRETS_FILE = 'client_secret.json'

# Directory to store user's tokens
TOKEN_STORAGE_DIR = 'tokens'
TOKEN_FILE = os.path.join(TOKEN_STORAGE_DIR, 'token.json')

class GoogleOAuthService:
    def __init__(self, update_notification_callback=None):
        self.credentials = None
        self.update_notification_callback = update_notification_callback
        os.makedirs(TOKEN_STORAGE_DIR, exist_ok=True) # Ensure token directory exists

    def _update_notification(self, message):
        if self.update_notification_callback:
            self.update_notification_callback(message)
        else:
            print(message)

    def authorize(self):
        """
        Authorizes the user and returns Credentials.
        If a refresh token is available, it will be used to refresh the access token.
        Otherwise, it will initiate a new authorization flow, opening a browser window.
        """
        self.credentials = None

        # Load existing credentials if available
        if os.path.exists(TOKEN_FILE):
            try:
                self.credentials = Credentials.from_authorized_user_file(TOKEN_FILE, SCOPES)
                self._update_notification("Loaded existing credentials.")
            except Exception as e:
                self._update_notification(f"Error loading existing credentials: {e}. Re-authorizing.")
                self.credentials = None # Invalidate credentials if loading fails

        # If credentials are not valid or don't exist, initiate the flow
        if not self.credentials or not self.credentials.valid:
            if self.credentials and self.credentials.expired and self.credentials.refresh_token:
                self._update_notification("Refreshing access token...")
                try:
                    self.credentials.refresh(Request())
                    self._update_notification("Access token refreshed.")
                except Exception as e:
                    self._update_notification(f"Error refreshing token: {e}. Re-authorizing.")
                    self.credentials = None # Invalidate if refresh fails
            
            if not self.credentials or not self.credentials.valid:
                # Flow for desktop application
                self._update_notification("Initiating new OAuth authorization flow...")
                flow = InstalledAppFlow.from_client_secrets_file(CLIENT_SECRETS_FILE, SCOPES)
                self.credentials = flow.run_local_server(port=0) # port=0 lets OS pick a free port
                self._update_notification("Authorization successful.")

        # Save credentials for the next run and print authorized email
        if self.credentials and self.credentials.valid:
            if self.credentials.id_token:
                try:
                    import jwt
                    decoded_token = jwt.decode(self.credentials.id_token, options={"verify_signature": False})
                    user_email = decoded_token.get('email')
                    if user_email:
                        self._update_notification(f"Authorized as: {user_email}")
                    else:
                        self._update_notification("Authorized, but email not found in token.")
                except ImportError:
                    self._update_notification("Authorized. Install 'PyJWT' for email logging if needed.")
                except Exception as e:
                    self._update_notification(f"Error decoding ID token: {e}")
            else:
                self._update_notification("Authorized. No ID token available to get email.")

            with open(TOKEN_FILE, 'w') as token:
                token.write(self.credentials.to_json())
            self._update_notification("Credentials saved.")
        
        return self.credentials

    def get_credentials(self):
        return self.credentials
